apps/expo-go/metro.config.js

```diff
@@ -14,25 +14,61 @@ config.watchFolders = [
-  // Assuming home only serves for Expo Go,
-  // the strategy here is to serve react-native imports from `react-native-lab/react-native` but not its transitive dependencies.
-  // That is not ideal but should work for most cases if the two react-native versions do not have too much difference.
-  // For example, `react-native-lab/react-native/node_modules/@react-native/polyfills` and `node_modules/@react-native/polyfills` may be different,
-  // the metro config will use the transitive dependency from `node_modules/@react-native/polyfills`.
-  /\breact-native-lab\/react-native\/node_modules\b/,
-
-  // Copied from expo-yarn-workspaces
-  /\/__tests__\//,
-  /\/android\/React(Android|Common)\//,
-  /\/versioned-react-native\//,
-];
+// Disable Babel's RC lookup, reducing the config loading in Babel - resulting in faster bootup for transformations
+config.transformer.enableBabelRCLookup = false;
+
+module.exports = withReactNativeLabPackages(config);
+
+/**
+ * Configure Metro to resolve core React Native packages from the react-native-lab submodule.
+ * Expo Go requires specific fixes which are shipped through this submodule.
+ * This will override and not follow Node module resolution to ensure single package versions are bundled.
+ *
+ * @param {import('expo/metro-config').MetroConfig} config
+ * @returns {import('expo/metro-config').MetroConfig}
+ */
+function withReactNativeLabPackages(config) {
+  // Keep this list up to date for every React Native upgrade.
+  // These packages are all dependent on the current version in react-native-lab.
+  const linkedPackages = [
+    'react',
+    'react-native',
+    // '@react-native/assets-registry', - This is overwritten from Expo CLI and replaced by a virtual module
+    // '@react-native/codegen', - This is not a library used within the app bundle
+    // '@react-native/js-polyfills', - This is not a library used within the app bundle, but configured through `config.serializer.getPolyfills`
+    '@react-native/normalize-colors',
+    '@react-native/virtualized-lists',
+  ];
+
+  // The react-native-lab location to use when redirecting resolutions to react-native-lab
+  const reactNativeLabsPath = path.join(monorepoRoot, 'react-native-lab/react-native/node_modules');
+
+  // Allow Metro to resolve from react-native-lab dependencies and packages
+  config.watchFolders.push(path.join(reactNativeLabsPath, '..'));
+
+  // Ensure the polyfills are loaded from react-native-lab
+  config.serializer.getPolyfills = () =>
+    require(require.resolve('@react-native/js-polyfills', { paths: [reactNativeLabsPath] }))();
+
+  // The import matcher that matches any of the list packages, e.g. `<packageName>` or `<packageName>/nested/file`
+  const importMatcher = new RegExp(`^(?:(${linkedPackages.join('|')}))(?:(/|$))`);
+
+  // Rewrite imports to the react-native-lab linked packages when on Android or iOS
+  config.resolver.resolveRequest = (context, moduleImport, platform) => {
```

Comment:
nice custom resolver. we used to test `apps/native-component-list` and `apps/test-suite` when running expo-go qa. would you like to replicate the same logic to ncl and test-suite?

maybe we can dynamically pull the `withReactNativeLabPackages` when `process.env.EXPO_GO === true` on ncl and test-suite.

Reply:
I'm not sure if I follow, I thought the only place where we needed `react-native-lab`'s version of React Native, was Expo Go. Why would we need to use this for NCL and Test Suit too?

Reply:
good point. the case happens when expo-go (react-native-lab) has an older version than node_modules/react-native. in the meantime, we cannot use expo-go to test NCL and TestSuite. otherwise people have to add resolutions to pin down react-native version before testing.

as mentioned previously, that's not a big issue. we could remove the react-native-lab resolution code in NCL and TestSuite first. and add back workaround when the error happens again.

Reply:
As we pair-programmed together, only a more modern package manager can help us in this scenario. We can't just hotswap a single package, without it's dependants into the bundle, that would only create more problems.

Dropped any references to `react-native-lab` from the other `metro.config.js` instances. We can circle back to this once we have a more modern package manager in `expo/expo`.